src/scripts/controllers/TasksListController.js

```diff
@@ -1,14 +1,57 @@
 import TasksList from '../models/TasksList.js';
+import StorageService from '../services/StorageService.js';
 import TasksListView from '../views/TasksListView.js';
 
 export default class TasksListController {
-  constructor(tasksListContainerId, initialTasks) {
+  constructor(tasksListContainerId, tasksStorageKey) {
     this.tasksListContainerId = tasksListContainerId;
-    this.tasksList = new TasksList(initialTasks);
-    this.tasksListView = new TasksListView(this.tasksList.getTasks);
+    this.storageService = new StorageService(tasksStorageKey);
+    this.tasksList = new TasksList(this.storageService.getData());
+
+    const eventHandlers = {
+      handleToggleBtn: this.handleToggleIsCompleted,
+      handleEditTask: this.handleEditTask,
+      handleRemoveTask: this.handleRemoveTask,
+    };
+    this.tasksListView = new TasksListView(this.tasksList.getTasks, eventHandlers);
   }
 
   build = () => {
     this.tasksListView.render(this.tasksListContainerId);
+  };
+
+  updateStorage = () => {
+    this.storageService.storeData(this.tasksList.getTasks);
+  }
+
+  addNewTask = (description) => {
+    const newTask = this.tasksList.createAndAddTask({
+      description,
+      isCompleted: false,
+    });
+    this.tasksListView.renderNewTask(newTask);
+    this.updateStorage();
+  };
+
+  handleToggleIsCompleted = (taskId) => {
+    const newIsCompleted = this.tasksList.toggleTaskIsCompleted(taskId);
+    this.tasksListView.updateTaskIsCompleted(taskId, newIsCompleted);
+    this.updateStorage();
+  };
+
+  handleEditTask = (taskId, newDescription) => {
+    if (newDescription) {
+      this.tasksList.setTaskDescription(taskId, newDescription);
+      this.tasksListView.handleTaskEditingState(taskId);
+      this.updateStorage();
+    } else {
+      this.handleRemoveTask(taskId);
+    }
+  }
+
+  handleRemoveTask = (taskId) => {
+    this.tasksList.removeTaskById(taskId);
+    this.tasksListView.removeTaskFromScreen(taskId);
+    this.updateStorage();
   }
 }
```

Comment:
Deleting a task should update all remaining items' indexes, so they represent the current list order and are unique. Currently, your indexes are not updating when an item is deleted.

- in the example below I added 6 items, then I deleted the 2nd and 5th items and this is how your indexes look like:

![image](https://user-images.githubusercontent.com/38088818/171918023-ab204c34-2ae9-49da-b805-c94726f14027.png)

Reply:
## Hi @KDlamini 👋,

**Thank you so much for the feedback!** About the required change of updating the tasks ``indexes`` when the user deletes a task, I'm aware this was one of the requirements, but sometimes as a software engineer is our job to try to understand the ``why`` instead of just following the requirements blindly. 

The way I see it, the reason for updating the tasks indexes is:
- to always render the task list in order
- to have always a unique ``index/id`` for each task

**My solution does that!** Even when adding new tasks or deleting an existing one, the tasks ``ids/indexes`` represents the current list order (there are 'gaps' between the id's numbers but they are all in order) and are unique for each task. 

The reason I did this way instead of the way it was suggested in the requirements is that iterating through an array can be a VERY expensive operation and we should, in my opinion, avoid doing that if it's unnecessary. 
In this case, for example, I believe it's totally unnecessary. This won't give me any new functionality, this won't improve my code quality, and this won't give me ANY benefit over my solution at all.  To be honest, this will actually make my code less efficient than it is right now... 
Imagine if a had a tasks list with ``100,000 tasks`` in it. In this case, when the user deletes the first task, for example, I would have to update the index of ``99,999 tasks``. That's a very expensive and inefficient operation. This is the worst-case scenario, I know, but in the average case scenario where the user deletes ``the task in the middle with the index of #50,000``, I would still have to update ``49,999 tasks``! And the worst thing is I'm iterating through this huge array EVERY TIME the user deletes a task!

**Note:** my solution for the tasks ``ids/indexes`` also allows me to implement the dragging feature with the same efficiency that if I had implemented the solution it was suggested in the requirements.

Being that said, I can refactor my code to use the requirement's solution, but **I strongly doubt this is the best way to address this problem** (I'm not saying my solution is the best approach either, but it is definitely better than the one it was suggested in the requirements).